### ISSUE TITLE
Non-interactive tunnel setup using CLI (for dockerized code-server)

### ISSUE THREAD

**Issue by mishamsk** [microsoft/vscode#170013]
Hi,

The insider's code-server CLI had a "serve-local" option which allowed the server to be baked in a docker image and recreated on demand (albeit only supporting local access via IP, no discovery via GitHub.dev).

The new CLI doesn't support that and forces a user to interactively authenticate against a provider. I can see that `code tunnel user` has a subcommand `login` with `--access-token` argument, which I hoped would solve the issue (and even allow using GitHub.dev for remote discovery of the tunnels). However, despite logging in did work with my personal access token, when I tried running the tunnel - I got 401 error.
![CleanShot 2022-12-25 at 15 53 18@2x](https://user-images.githubusercontent.com/5206955/209481712-27aaac45-ad74-434e-b131-e7bf1cbe5c44.png)

I could not find a documentation how to pre-authorizes a machine and get a valid access token. Having this or other alternative option to authenticate code server would be great.

**Comment by connor4312** [comment]
`--access-token` currently only support Github OAuth refresh tokens, not PATs

**Comment by mishamsk** [comment]
@connor4312 thanks. I'll try that, assuming refresh tokens do not expire, this seems like a reasonable workaround.

Thought I'll elaborate why I was asking in the first place: I run the server on Sinology NAS. Unfortunately it won't run natively due to library compatibility, but it works alright in docker. I have quite a few containers all managed under a single compose file, which I update from time-to-time, spinning the project up & down => without a way to pre-auth vs code server, I have to manually spin up the server container and do re-auth. The server is used mostly to enable coding on iPad

**Comment by fd** [comment]
Running into the same issue. @mishamsk, did you get this working?

**Comment by mishamsk** [comment]
@fd nope. haven't been using my server lately at all, glued to my laptop for the last 6 moths 24/7...

**Comment by fd** [comment]
> `--access-token` currently only support Github OAuth refresh tokens, not PATs

@connor4312 Well, I can confirm it definitely does not to take a `ghr_` refresh token. Also this simply does not make sense as the [login procedure](https://github.com/microsoft/vscode/blob/37577f6b29f6f0b311d4db4c762dee51bbbdfeca/cli/src/auth.rs#L386-L408) just chugs the access token in the `StoredCredential.access_token` field and the stores those credentials.

This feature seems to be completely broken. I've tried with many different github token types, oauth scopes.

**Comment by tasdevani21** [comment]
> It seems tightly coupled to the _VS Code_ Github Oauth App. When going through the device login flow with the client id from the code cli seems to make it work.

Can you share more details on this? Did you have to setup a github OAuth App? And how can we get the client id from the code cli? any instructions that you are able to share? Thank you!

**Comment by fd** [comment]
No need for a Github Oauth App. Only the the oauth device flow can be used (with the client id from the code cli). The client id is in the source code for the new code cli.

**Comment by jjcasmar** [comment]
Did you manage to get this working? I am trying to automate the creation of a tunnel when I start a docker container

**Comment by connor4312** [comment]
This can now be done with the following steps:

1. [Install the dev tunnels CLI](https://learn.microsoft.com/en-us/azure/developer/dev-tunnels/get-started?tabs=macos#install)
1. Run `devtunnel login` (see --help for possible options)
1. Run `devtunnel token <tunnel id> --scope manage --scope host`
1. Run `code-insiders tunnel --tunnel-id <tunnel id> --host-token <token> --name some-tunnel-name`
1. https://insiders.vscode.dev/tunnel/your-tunnel-name

**However**, note that devcontainers now supports connecting to a container over tunnels. It will be supported on vscode.dev as well soon.

**Comment by jjcasmar** [comment]
@connor4312 `devtunnel token <tunnel id> --scope manage --scope host` gives an error: Tunnel not found

**Comment by connor4312** [comment]
Sorry, missed a step, you need to `devtunnel create` first

**Comment by jjcasmar** [comment]
That sequence of commands works. But I think its not what we were asking in this issue. This is still interactive, as we have to log in the first time. What we were asking is the possibility to create a personal token in github and use that token to log in when creating the tunnel. So we could do something like `code tunnel --access-token <token given in github ui>`
